book/src/leader-rotation.md

```diff
@@ -37,21 +42,20 @@ from slot 200 until it is updated.
 No skews can exist because *everyone* that is voting with the network has
 skipped 100 and 101 when their root reaches 102.
 
-### Leader Schedule Rotation Across Epochs
+### Leader Schedule Rotation with Epoch-Scale partitions.
 
-If the next slot skips an epoch, it is due to a considerable network failure,
-and the leader schedule from the previous epoch is still valid until the root
-fork is updated.
+Consider the following scenario.  Two partitions that are generating half of the
+blocks each.  Neither is coming to a definitive supermajority branch.  Both will
+cross epoch 100 and 200 without actually committing to a root and therefore a
+new Leader Schedule.
 
-For example:
+In this unstable scenario multiple valid leader schedules exist.
 
-1. Large network timeout occurs.
-2. Vote occurs on fork 201
-3. Root checkpoint moves from 99 to 102
-4. Leader scheduler is updated for epoch 300+, using fork 102.
+* An LS is generated for every fork whose direct parent is in the previous
```

Comment:
I don't understand how an LS could be valid without a root, I'm also a little wary about how a node gets forks so high without voting.  wouldn't every node have to vote, thereby arriving at a root?

Reply:
@rob-solana what do we do if there is a 50/50 partition that lasts longer than an epoch?

Reply:
Once you start getting divergent leader schedules, that's when the number of possibilities for forks starts growing exponentially. Is the hope that our fork choice rules resolve these eventually, or at least from a single node's perspective, that after max lockout we can prune/ignore many of these possibilities?

Reply:
@carllin 

Yea, it doubles every epoch.  But, the probability of a epoch scale partition should be 0.   

Two networks, split at 50/50, continue indefinitely.  Neither generating a majority, so votes keep expiring.  But the blocks are connected, and some of the blocks cross the epoch boundary.  The block that crosses the epoch boundary is not aware if it will ever be the root or not.

Reply:
@rob-solana idea of a "root" is not network wide.  its local to the client.  All this is saying is that a block referencing a fork that transitioned a epoch boundary 1 epoch ago should use the new schedule.

Reply:
@rob-solana @carllin added another option

Reply:
I recognize that a root is a local concept.

if this is a toss-up, I'd prefer to stay with the old LS until the partition clears up.

Reply:
@rob-solana, how do you stay with it?  I guess the case I am thinking of, can you propose a block connecting to a block which you didn’t vote for?

Reply:
yeah, why not?

Reply:
@rob-solana, so in that case the proposed block may be attached to epoch block, and there is no guarantee that it is a root.

Fundamentally, if we pick A over C, we can only hope that if the epoch is long enough that the network doesn’t partition.  

I was thinking the epoch would be two weeks to really limit any kind of scheduling censorship.

Reply:
I'm having trouble imagining how we'd get to a partition lasting longer than an epoch even with an epoch that's only 128 * 32 blocks (like an hour), but the longer the better, I guess.  We will depend on stable schedules (or another kind of scheduling).

Reply:
2 week epochs give humans time to react and do things IRL to fix if required

Reply:
@rob-solana Yea, you are correct.  An epoch that is greater than the median partition time + 6 standard deviations would mean that the likelihood of a partition lasting longer than an epoch is 1 in 1 million.

Reply:
Also, @rob-solana 128 * 32 blocks would generate 4096 - 32 roots.

Reply:
> Also, @rob-solana 128 * 32 blocks would generate 4096 - 32 roots.

@aeyakovenko assuming nodes are voting at a greedy cadence, but if nodes hold their votes waiting for a majority or some amount of stake, there could be fewer roots.  My pick of 128*MAX_LOCKOUT was intended to be large enough that voting would have to have taken place.

Reply:
I'm back to worrying that root skew will mess up schedule generation...

Reply:
how?  the algorithm isn't based on root.  Any block derived from the epoch crossing block is computed with the updated schedule after the offset.

Reply:
Maybe I'm confused.  Here's the scenario I'm thinking of:

100 block epochs.  Network with only nodes A and B staked.

A and B chugging along make roots at blocks 102 and 103, respectively (because they have different voting strategies).

They each run generate_schedule() for the epoch starting at 300 at their respective roots, but block 103 includes a new stake for node C.

Reply:
@rob-solana

99 <- 102 <- 103?

Block 102 is the block that crossed the epoch boundary, so it's the block that is used for calculating the leader schedule.  Regardless if the node's root is 102 or 103 or any other block.

We can just delay that calculation until a root that crosses the epoch boundary is selected.